lib/Doctrine/ORM/Query/QueryExpressionVisitor.php

```diff
@@ -74,7 +74,7 @@ public function __construct($queryAliases)
      * Gets bound parameters.
      * Filled after {@link dispach()}.
      *
-     * @return \Doctrine\Common\Collections\Collection
+     * @return ArrayCollection
```

Comment:
Isn't better to keep the Interface as return type ?

Reply:
I'm not sure, let see what other people say.

Reply:
People will code against your advertised return type. The question is: do you need this method to return an ArrayCollection for it to be used properly, or would any Collection be the same, meaning the choice of ArrayCollection is an implementation detail internal to the method itself?

Reply:
People will code against your advertised return type. The question is: do you need this method to return an ArrayCollection for it to be used properly, or would any Collection be the same, meaning the choice of ArrayCollection is an implementation detail internal to the method itself?

Reply:
I agree with @kuraobi

Reply:
I don't _need_ it to be advertised as such, but as a bonus, people can have methods specific to ArrayCollection show up in their auto-completion. Also, their SA tools will not complain if they do use such methods.

Reply:
Sure, but the cost is that you are committing to return an ArrayCollection until at least the next major release, so you have to be sure that you're ok with that. If you're fine with the trade-off, go for it.

Reply:
@kuraobi @ostrolucky would you consider suddenly returning something else than `ArrayCollection` here a BC-break?

Reply:
It's BC breaking and irrelevant change for goal of this PR. We shouldn't be changing public API in this PR, that's not its goal. Previously, API said this method needs to return Collection, now we say it needs concrete implementation of Collection.

Reply:
For people using SA, @return ArrayCollection is the same as `function f(): ArrayCollection`, because their code relies on that. So changing the return type would be a BC break IMHO.

Reply:
No I mean changing the actual returned type:

```diff
/** @return Collection */
function ()
{
-    return ArrayCollection();
+    return SomeOtherCollection();
}
```

If this is a BC-break, then we are already commited to return `ArrayCollection`. If this isn't, I should revert this change.

Reply:
It's a LSP breaking change. I don't know why are we considering this change here, API changes are not a goal of this PR. Return type was defined here and was correct. Did Psalm do it automatically? I would report it as a psalm bug.

Reply:
No, your last diff isn't a bc break

Reply:
Ok I will revert this. Maybe the right change would be to have this:

```
/**
 * @return Collection
 * @psalm-return ArrayCollection
 */

Reply:
I don't think so, psalm-return overrides return. What we are trying to tell consumers here is that they cannot rely on us returning ArrayCollection, we can change it any time to different instance of Collection.

Reply:
ArrayCollection implements both Collection and Selectable, so it might be a better typehint. However @ostrolucky is right that it could theoretically return a PersistentCollection as well (it does not yet I believe)

Reply:
I could use `Collection&Selectable`, WDYT?

Reply:
It's still narrowing the type, which is a BC break. Please propose it for master if you want that

Reply:
Narrowing a return type in a concrete class is not a BC break AFAIK, unless you have children classes and no interface to enforce a common contract.

Reply:
Well there is no interface here, and the class is not final…

Reply:
```
>>> interface a{function a():IteratorAggregate;}
>>> class b implements a{function a():ArrayObject {}}
>>> class c extends b {function a(): IteratorAggregate {}}
PHP Fatal error:  Declaration of c::a(): IteratorAggregate must be compatible with b::a(): ArrayObject
```

Reply:
Yeah so we're all saying the exact same thing :laughing: